src/dns_data.py

```diff
@@ -64,50 +50,36 @@ def create_dns_record_provider_data(
     return DNSRecordProviderData(dns_entries=statuses)
 
 
-def has_a_zone_changed(
+def has_changed(
     relation_data: list[tuple[DNSRecordRequirerData, DNSRecordProviderData]],
     topology: models.Topology | None,
+    last_valid_state: dict[str, typing.Any],
 ) -> bool:
-    """Check if a zone definition has changed.
+    """Check if the dns data has changed.
 
-    The zone definition is checked from a custom hash constructed from its DNS records.
+    This could be a change in a zone, or a removal/addition of a zone,
+    or a change in the topology.
+    We use the state.json file to compare the state when
+    the last configuration was minted to the current one.
 
     Args:
         relation_data: input relation data
         topology: Topology of the current deployment
+        last_valid_state: The last valid state, deserialized from a state.json file
 
     Returns:
         True if a zone has changed, False otherwise.
     """
     zones = dns_record_relations_data_to_zones(relation_data)
-    logger.debug("Zones: %s", [z.domain for z in zones])
-    for zone in zones:
-        try:
-            zonefile_content = pathlib.Path(
-                constants.DNS_CONFIG_DIR, f"db.{zone.domain}"
-            ).read_text(encoding="utf-8")
-            metadata = _get_zonefile_metadata(zonefile_content)
-        except (
-            InvalidZoneFileMetadataError,
-            EmptyZoneFileMetadataError,
-            FileNotFoundError,
-        ):
-            return True
-        if "HASH" in metadata and hash(zone) != int(metadata["HASH"]):
-            logger.debug("Config hash has changed !")
-            return True
 
-    if topology is not None and topology.is_current_unit_active:
-        try:
-            named_conf_content = pathlib.Path(
-                constants.DNS_CONFIG_DIR, "named.conf.local"
-            ).read_text(encoding="utf-8")
-        except FileNotFoundError:
+    if "zones" not in last_valid_state or last_valid_state["zones"] != zones:
+        return True
+
+    if topology is not None and "topology" in last_valid_state:
```

Comment:
I believe you are missing
```
if topology is None and "topology" in last_valid_state:
    return True
```

Reply:
I revisited that and reworked it to be way simpler.